src/Operations.jl

```diff
@@ -596,8 +596,10 @@ function collect_artifacts(pkg_root::String; platform::AbstractPlatform=HostPlat
 
             # If there is a dynamic artifact selector, run that in an appropriate sandbox to select artifacts
             if isfile(selector_path)
-                select_cmd = Cmd(`$(gen_build_code(selector_path)) $(triplet(platform))`)
-                addenv(select_cmd, "JULIA_LOAD_PATH" => "@stdlib")
+                # Despite the fact that we inherit the project, since the in-memory manifest
+                # has not been updated yet, if we try to load any dependencies, it may fail.
+                # Therefore, this project inheritance is really only for Preferences, not dependencies.
```

Comment:
I must say I don't understand what this comments is meant to convey.

Reply:
https://github.com/JuliaLang/Pkg.jl/pull/2024 added the ability for packages to give a little snippet of code that should be run in order to help `Pkg` determine which artifact to install at `Pkg.add()` time.  The canonical example is CUDA version (e.g. CUDA 9 vs CUDA 10); that's not something Julia natively knows about (as opposed to processor architecture, e.g. `x86_64` vs. `aarch64`).  The eventual design was to have a script that prints out TOML chunks that represent the artifacts that need to be downloaded.

Valentin and Simon have been working on a way to improve the MPI situation, and their current thinking is to use `Preferences` to allow setting a particular MPI backend, which then is used to coordinate which artifacts get downloaded by downstream MPI-aware packages.  The mechanism by which this happens is that the `select_artifacts.jl` hook inspects the preferences of the current environment, and uses that to inform the artifact selection machinery accordingly.

We run the `select_artifacts.jl` file in a new Julia process and without any particular project, using the same machinery as `Pkg.build()`.  Because Preferences reading relies on the `Base.LOAD_PATH` stack, it's important to pass `--project=$(Base.active_project())` to the Julia subprocess so that it sees the same preferences as when you are actually running the package.  However, because this is happening in the middle of `Pkg.add()`, this comment is meant to educate nosy users that despite the fact that we're running with a project that claims to have a bunch of dependencies, those dependencies may not be loadable via `using Foo`, because we haven't finished the `Pkg.add()` operation yet.